Tensile/Utils.py

```diff
@@ -22,30 +22,137 @@
+        sys.stdout.write('\r')
+        sys.stdout.write(' ' * (len(self.message) + self.maxTicks + 15) + '\r')
+
+        stopTime = time.time()
+
+        sys.stdout.write(f"{self.message}... Done in {stopTime - self.createTime:.1f} secs\n")
+        sys.stdout.flush()
 
 class SpinnyThing:
-    def __init__(self):
-        self.chars = ['|', '/', '-', '\\']
-        self.index = 0;
+    """A class to display a spinning indicator in the console for long-running operations.
+
+    This class provides a simple way to visually indicate that a long-running operation
+    is in progress by displaying a spinning character in the console. The spinner is
+    updated at regular intervals, and a completion message with the elapsed time is
+    displayed when the operation finishes.
+
+    Attributes:
+        msg: The message displayed alongside the spinner.
+        chars: The sequence of characters used for the spinner animation.
+        index: The current index in the `chars` list for the spinner.
+        count: A counter to control the update frequency of the spinner.
+        createTime: The timestamp when the spinner was created.
+    """
+    def __init__(self, msg: str):
+        self.msg: str = "# " + msg
+        self.chars: List[str] = ['|', '/', '-', '\\']
+        self.index: int = 0
+        self.count: int = 0
+        self.createTime: float = time.time()
+
+    def increment(self):
+        """Increments the spinner's position and updates the display if necessary."""
+        self.count += 1
+        if self.count % 3 != 0:
+            return
+
+        sys.stdout.write('\r' + ' ' * (len(self.msg) + 10))
+        sys.stdout.flush()
 
-    def increment(self, value=1):
-        sys.stdout.write('\b' + self.chars[self.index])
+        sys.stdout.write('\r' + self.msg + " " + self.chars[self.index])
         sys.stdout.flush()
         self.index = (self.index + 1) % len(self.chars)
 
     def finish(self):
-        sys.stdout.write('\b*\n')
+        """Clears the spinner and displays a completion message with the elapsed time."""
+        sys.stdout.write('\r' + ' ' * (len(self.msg) + 10))
+        sys.stdout.flush()
+
+        stopTime = time.time()
+        elapsedTime = stopTime - self.createTime
+
+        sys.stdout.write('\r' + self.msg + f'... Done in {elapsedTime:.1f} secs\n')
         sys.stdout.flush()
 
 def iterate_progress(obj, *args, **kwargs):
-    try:
-        progress = ProgressBar(len(obj))
-    except TypeError:
-        progress = SpinnyThing()
+    if 'desc' not in kwargs:
+        printWarning("No message provided for TQDM progress bar/spinner", DeveloperWarning)
+        kwargs['desc'] = 'Processing unknown function'
+    if 'total' in kwargs:
+        progress = ProgressBar(kwargs['total'], kwargs['desc'])
+    else:
+        try:
+            progress =  ProgressBar(len(obj), kwargs['desc'])
+        except TypeError:
+            progress = SpinnyThing(kwargs['desc'])
```

Comment:
It appears we now have the ability to consistently use the progress bar - am I correct?

Reply:
We do now have the ability to every call to `Utils.tqdm` to use a progress bar. Either the object has to have `__len__` defined on it or you can pass in the kwarg `total=N` where N is the known length. If neither of these are true, it will default to the Spinner.